### PR TITLE
ENH vs2017

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [ ] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge-admin rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,59 +1,84 @@
 # version numbering is explained at https://stackoverflow.com/questions/42730478/version-numbers-for-visual-studio-2017-boost-and-cmake
 #     This version number follows the "platform toolset"
-{% set vcver="14" %}
-{% set vsyear="2015" %}
-{% set fullver="14.0.25420" %}
+{% set vcver="14.1" %}
+{% set vsyear="2017" %}
+{% set fullver="14.16"%}
```
Comment (isuruf) [inline recipe/meta.yaml]: Use `14.16.27012` here?

Comment (beckermr) [reply recipe/meta.yaml]: yup! I just noticed that when looking at the CI logs

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,59 +1,84 @@
...
     requirements:
       run:
         # Set lower bound for runtime (no upper bound).  Upper bound is still present, but
         #    it's part of the run_exports for the compiler: assume that when vcver's major version changes,
         #    MSFT has broken our binary compatibility.
-        - {{ pin_subpackage('vs' ~ vsyear ~ '_runtime', max_pin=None) }}
+        - {{ pin_subpackage('vs' ~ runtime_year ~ '_runtime', max_pin=None) }}
     about:
       home: https://github.com/conda/conda/wiki/VC-features
       license: Modified BSD License (3-clause)
```
Comment (isuruf) [inline recipe/meta.yaml]:
```suggestion
      license: BSD-3-Clause
```

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge/vc @conda-forge/core @isuruf @jjhelmus This one is ready for review!

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,62 +1,95 @@
...
+#    mixed up with the runtime from VS2015 - you do actually need VS2017 runtime for things compiled with
+#    VS2017, and the runtime we have here is backwards-compatible with things compiled with VS2015.
+{% set runtime_year = "2015" %}
 
 package:
   name: vs{{ vsyear }}
   version: {{ fullver }}
 
 build:
-  number: 11
```
Comment (jjhelmus) [inline recipe/meta.yaml]: This could be reset to 0 since the package name a version changed.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,62 +1,95 @@
...
+        - vc{{ vcver.split(".")[0] }}
+      run_exports:
+        strong:
+          # compatible within major version.  This is MSFT's incrementing of the UCRT version.  It was
+          #    14.0 with VS2015, and 14.1 with VS2017 - if the major version changes, we're assuming
+          #    that binary compatibility breaks.  This has a minimum bound equal to vcver, so building with
+          #    vs2017 here will require the vs2017-era runtime (the package is named vs2015_runtime, but the
+          #    version differs)
+          - {{ pin_subpackage('vc') }}
+          - {{ pin_subpackage('vs' + runtime_year + '_runtime') }}
```
Comment (beckermr) [inline recipe/meta.yaml]: This used to be `max_pin=None`. Should it go back?

File: recipe/meta.yaml (hunk 3 above)
Comment (beckermr) [reply recipe/meta.yaml]: done

**Comment by beckermr** [comment]
ok @conda-forge/vc all of the hacking here is done. Can we resolve my question above and get this merged?

File: recipe/install_runtime.bat [hunk 5]
```diff
@@ -1,30 +1,54 @@
...
+    echo No VS installation detected by vswhere.  Exiting.
+    exit 1
 )
 
+set "REDIST_ROOT=%BT_ROOT%\VC\Redist\MSVC\%runtime_version%\%VC_PATH%"
+echo "%REDIST_ROOT%"
+
+IF NOT EXIST "%BT_ROOT%\VC\Redist\MSVC\%runtime_version%" (
+   echo ============================================================================================
+   set "out=!out! Expected vcruntime140.dll version (from conda_build_config.yaml) was %runtime_version%."
```
Comment (isuruf) [inline recipe/install_runtime.bat]:
```suggestion
   set "out=!out! Expected vcruntime141.dll version (from conda_build_config.yaml) was %runtime_version%."
```

File: recipe/conda_build_config.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+update_version:
+  - 9
+# this is the version number reported by cl.exe
+cl_version:
+  - 19.16.27032.1
+# this version follows the version number reported for the vcruntim140.dll file.
+#     This will be in a folder like C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Redist\MSVC\14.15.26706\x64\Microsoft.VC141.CRT
+#     Note the 14.15.26706 in there - that's what we want, but in case their folder scheme changes, it's better to base it on the version of the file itself.
+runtime_version:
+  - 14.16.27012
```
Comment (isuruf) [inline recipe/conda_build_config.yaml]: Shouldn't this be `14.16.27032` if `cl_version` is `19.16.27032`?

File: recipe/meta.yaml (hunk 4 above)
Comment (beckermr) [reply recipe/meta.yaml]: bump on this run exports comment too

File: recipe/conda_build_config.yaml (hunk 6 above)
Comment (beckermr) [reply recipe/conda_build_config.yaml]: I have no idea. :) So if you think so, yes!

File: recipe/meta.yaml (hunk 4 above)
Comment (isuruf) [reply recipe/meta.yaml]: Yes. I guess they'll break compatibility in 15

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,62 +1,95 @@
 # version numbering is explained at https://stackoverflow.com/questions/42730478/version-numbers-for-visual-studio-2017-boost-and-cmake
 #     This version number follows the "platform toolset"
-{% set vcver="14" %}
-{% set vsyear="2015" %}
-{% set fullver="14.0.25420" %}
+{% set vcver = "14.1" %}
+{% set vsyear = "2017" %}
+{% set fullver = "14.16.27012" %}
```
Comment (isuruf) [inline recipe/meta.yaml]:
```suggestion
{% set fullver = runtime_version %}
```

Comment (beckermr) [reply recipe/meta.yaml]: We might need os.environ here

Comment (beckermr) [reply recipe/meta.yaml]: wait no

**Comment by beckermr** [comment]
Ok @isuruf this is ready to go!

**Comment by beckermr** [comment]
We good to merge here @isuruf ?

**Pull request merged**